### PR TITLE
Feature/web services enhancement diameter impelmentation

### PR DESCRIPTION
**Pull request by amitamir**
Thank you for your contribution.
Before submitting this PR, please make sure:
- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Help Libray document is updated accordingly for Feature changes.

### CODE REVIEW

File: Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs [hunk 1]
```diff
@@ -0,0 +1,2302 @@
...
+        }
+        private static string CreateAVPAsString(DiameterAVP avp, int identLevel = 1)
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            string identation = new string('\t', identLevel);
+            if (avp != null)
+            {
+                if (avp.DataType == eDiameterAvpDataType.Grouped)
+                {
+                    stringBuilder.Append($"{identation}<grouped avp name=\"{avp.Name}\" mandatory=\"{avp.IsMandatory.ToString().ToLower()}\">");
```
Comment (amitamir) [inline Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs:211]: remove space between grouped avp, set the ENUM value description for enumerated data types

File: Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs [hunk 2]
```diff
@@ -0,0 +1,2302 @@
...
+        {
+            mAct.RawResponseValues = ">>>>>>>>>>>>>>>>>>>>>>>>>>> REQUEST:" + Environment.NewLine + Environment.NewLine + RequestFileContent;
+            mAct.RawResponseValues += Environment.NewLine + Environment.NewLine;
+            mAct.RawResponseValues += ">>>>>>>>>>>>>>>>>>>>>>>>>>> RESPONSE:" + Environment.NewLine + Environment.NewLine + ResponseFileContent;
+            mAct.AddOrUpdateReturnParamActual("Raw Request: ", RequestFileContent);
+            mAct.AddOrUpdateReturnParamActual("Raw Response: ", ResponseFileContent);
+        }
+        private void ParseResponseAvpToOutputParams(DiameterAVP avp)
+        {
+            mAct.AddOrUpdateReturnParamActual($"AVP({avp.Name}): ", avp.Value);
```
Comment (amitamir) [inline Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs:2030]: 1. show parameter with only AVP name(without word AVP)
2.  in Path show AVPS/Name[num of instances]
3. if child, show in path the parent - AVPS/Parent(s)/..../Name[num of instances]

File: Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs [hunk 3]
```diff
@@ -624,7 +624,6 @@ public static void LoadUserMsgsPool()
             Reporter.UserMsgsPool.Add(eUserMsgKey.LoseChangesWarn, new UserMsg(eUserMsgType.WARN, "Save Changes", "The operation may result with lost of un-saved local changes." + Environment.NewLine + "Please make sure all changes were saved before continue." + Environment.NewLine + Environment.NewLine + "To perform the operation?", eUserMsgOption.YesNo, eUserMsgSelection.No));
 
             Reporter.UserMsgsPool.Add(eUserMsgKey.CompilationErrorOccured, new UserMsg(eUserMsgType.ERROR, "Compilation Error Occurred", "Compilation error occurred." + Environment.NewLine + "Error Details: " + Environment.NewLine + " '{0}'.", eUserMsgOption.OK, eUserMsgSelection.None));
-
```
Comment (amitamir) [inline Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs:627]: add the line back

File: Ginger/Ginger/Drivers/DriversConfigsEditPages/WebServicesDriverEditPage.xaml [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    <Page.Resources>
+        <BooleanToVisibilityConverter x:Key="B2V" />
+        <local:MultiBooleanToVisibilityConverter x:Key="MultiB2V"/>
+    </Page.Resources>
+
+
+    <Grid x:Name="xPageGrid" Background="{StaticResource $BackgroundColor_LightGray}" Grid.IsSharedSizeScope="True">
+        <DockPanel>
+            <Expander x:Name="xSettingsExpander" ExpandDirection="Down" IsExpanded="True" Style="{StaticResource $ExpanderStyle}" DockPanel.Dock="Top" Margin="0,0,0,0" BorderThickness="0">
```
Comment (amitamir) [inline Ginger/Ginger/Drivers/DriversConfigsEditPages/WebServicesDriverEditPage.xaml:20]: add spaces in tcp details, Advanced,equal the spaces in SoapUI Orch

File: Ginger/GingerCoreNET/DiameterLib/DiameterAVP.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            set
+            {
+                if (mVendorId != value)
+                {
+                    mVendorId = value;
+                    OnPropertyChanged(nameof(VendorId));
+                }
+            }
+        }
+        private bool mIsGrouped;
```
[the remark this answers is not in the document]
Comment (amitamir) [reply Ginger/GingerCoreNET/DiameterLib/DiameterAVP.cs]: removed isGrouped property as its unused

File: Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs (hunk 2 above)
Comment (amitamir) [reply Ginger/GingerCoreNET/DiameterLib/DiameterUtils.cs]: done

File: Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs (hunk 3 above)
Comment (amitamir) [reply Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs:627]: line was added back

**Review by MeniKadosh1 — APPROVED** [review]

**Pull request merged**